### ISSUE TITLE
List::Util's new mesh and zip are not compatible with List::SomeUtils equivalents 

### ISSUE THREAD

**Issue by oschwald** [houseabsolute/List-AllUtils#12]
After updating to the new `Scalar-List-Utils`, all of our uses of these via `List::AllUtils` broke. I am not sure what the best solution here is, perhaps not exporting the `List::Util`'s version or updating `List::SomeUtils` to match the `List::Util` behavior and doing a major release.

**Comment by autarch** [comment]
I think it's probably best to simply not export the List::Util versions. I think you could probably get those this way:

```
use List::AllUtils qw( uniq );
use List::Util qw( zip );
```

That somewhat defeats the purpose of AllUtils, but it's backwards compatible. It would be nice to add `zip_unflatten` to SomeUtils (without copying from MoreUtils because of license issues), and then AllUtils would offer everything you can get from the latest Util.

**Comment by petdance** [comment]
I can here to wave the flag about this.  We don't use List::AllUtils directly, but I did clone it into MyCompany::AllUtils, and a new List::Util 1.56 broke our code because of prototypes.

I'm changing the code in MyCompany::AllUtils to explicitly specify which functions to import from which module.

**Comment by oschwald** [comment]
We are now doing something similar. We have replaced `List::AllUtils` with our own module that has explicit imports.

**Comment by petdance** [comment]
Also interesting for those playing at home: Empty arrays will not trigger the prototype error in List::Util:

```
$ perl -MList::Util=zip -E'my @y; my @x; zip(@x,@y); say "happy"'
happy
$ perl -MList::Util=zip -E'my @y=(1); my @x=(3); zip(@x,@y); say "happy"'
Expected an ARRAY reference to zip at -e line 1.
```

**Comment by 0racle** [comment]
I was hit by this too. Considering this potentially breaks user code, List::AllUtils should be updated to prefer the SomeUtils version

**Comment by Grinnz** [comment]
> Also interesting for those playing at home: Empty arrays will not trigger the prototype error in List::Util:
> 
> ```
> $ perl -MList::Util=zip -E'my @y; my @x; zip(@x,@y); say "happy"'
> happy
> $ perl -MList::Util=zip -E'my @y=(1); my @x=(3); zip(@x,@y); say "happy"'
> Expected an ARRAY reference to zip at -e line 1.
> ```

That's because this function does not have a prototype, this is a parameter validation error. If you pass empty arrays there are no parameters to validate.

**Comment by Grinnz** [comment]
I don't see any forward-compatible option other than hardcoding where each function is imported from.

**Comment by petdance** [comment]
> That's because this function does not have a prototype, this is a parameter validation error. 

Indeed. I mentioned it because my first debugging attempts were passing empty arrays and I couldn't get it to fail. Oops.

**Comment by autarch** [comment]
This is fixed in v0.19. I simply hard-coded it to skip `mesh` and `zip` from `List::Util` for now.

I considered the idea of hard-coding the location of _every_ function, but I wasn't sure what to do when that list is out of date compared to the current install util modules. Right now this just works, and you get the new functions, with a defined order of what wins.

One possibility might be to hard-code a few more skips for things in `SomeUtils` that are the same in `Util`, and then _switch_ the ordering to prefer `SomeUtils` over `Util` in the future. That way if `Util` does something similar in the future, nothing breaks. Since I control `SomeUtils` I can be sure I will never add a function already in `Util`, so this seems safe.

How does that sound?

(note, I'm closing the issue since the bug is fixed, but we can continue the discussion here)
